### ISSUE TITLE
SC2218 not discovered when function combined with logical operators

### ISSUE THREAD

**Issue by olivergondza** [koalaman/shellcheck#3070]
First off, thanks a lot for maintaining this great project!

#### For bugs
- Rule Id (if any, e.g. SC2218):
- My shellcheck version (`shellcheck --version` or "online"): 0.10.0
- [x] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2218)
- [x] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/bash
false || run_fun

function run_fun() {
    echo "Fun run!"
}
```
When run, it fails with `line 2: run_fun: command not found`. The same happens for `true && run_fun`.

This can be tricky for the patterns of `potentially_failing_operation || diagnose_and_exit`. If the `diagnose_and_exit` is defined later, one will only learn that in unlikely circumstances (so the bug can exist undiscovered for a long time), and there is no diagnostics one hoped to acquire.

#### Here's what shellcheck currently says:

No error reported

#### Here's what I wanted or expected to see:

SC2218 (error): This function is only defined later. Move the definition up.

**Comment by koalaman** [comment]
This check from 2017 was very limited and naive to account for possibilities like `run_fun || run_fun() { backup implementation; }`, but thanks to ShellCheck's new CFG support it was simple to make it way more robust. It did indeed appear to find a number of unlikely circumstances where functions would fail to cleanup on error and such, just like you predicted.

Thanks!
